1105 Moved functional of Save into Disk and added test

yasamprom opened the pull request:
objectionary/eo#1105

yasamprom commented:
@Graur please check

Graur reviewed (approved):
@yasamprom Good one
@yegor256 Check it please

yasamprom commented:
@yegor256 Could you please see?

yasamprom commented:
@yegor could you please merge? It is needed for moving on doing the task

yegor256 commented:
@yasamprom there is an obvious code duplication between `Save` and `Disk`: making parent dir. Also, why did we move domain specific functionality (dir making) to `Disk`? `Disk` is supposed to be a proxy for `java.nio.Files`. Does `Files` create directory too?

yasamprom commented:
@yegor256 As we decided, I removed `Disk` and `Save` and added `Home` instead. Now we use it for modifying paths (in future) and saving our files.
Please check

yegor256 commented on eo-maven-plugin/src/main/java/org/eolang/maven/Home.java:
```diff
@@ -0,0 +1,246 @@
+     * @throws IOException if fails
+     */
+    public long write(final Input stream) throws IOException {
+        final File dir = this.cwd.toFile().getParentFile();
```
@yasamprom `cwd` means "current working directory". Why it's a file to write? I believe, `Home` has wrong design. Should be:

```
Home {
  Path cwd;
  void save(Path, String);
  String load(Path);
  String rel(Path);
}
```

yasamprom replied:
@yegor256 let's name it `path` instead of cwd? I don't see reasons to add second argument in `save` because we  don't have use cases where we need to have `cwd` and `file` separately.

yasamprom replied:
> @yegor256 let's name it `path` instead of `cwd`? I don't see reasons to add second argument in `save` because we don't have use cases where we need to have `cwd` and `file` separately.

And if we need it some day we will always be able to use `new Home(<needed path>)`

yasamprom replied:
> > @yegor256 let's name it `path` instead of `cwd`? I don't see reasons to add second argument in `save` because we don't have use cases where we need to have `cwd` and `file` separately.
> 
> We can always use `new Home(<needed path>)`

Well, then the `Home` name will be strange. I will add `path` argument to`save`

yegor256 replied:
@yasamprom `cwd` will help us print log messages nicely, because we will know where is the "home" and how relevant is the `path` to it

yegor256 commented:
@yasamprom see above

yasamprom commented:
@yegor256 Done

yasamprom commented:
@yegor256 Would you please see?

yegor256 commented on eo-maven-plugin/src/main/java/org/eolang/maven/Home.java:
```diff
@@ -0,0 +1,257 @@
+     * @throws IOException If fails
+     */
+    public long save(final Path path, final InputStream stream) throws IOException {
+        return this.write(path, new InputOf(stream));
```
@yasamprom you have a lot of code duplication in these methods. It's better to use existing methods, instead of calling `this.write` from each one. See how it's done in ctors of `Save`

yasamprom replied:
> @yasamprom you have a lot of code duplication in these methods. It's better to use existing methods, instead of calling `this.write` from each one. See how it's done in ctors of `Save`

@yegor256 So, we will use class like that: `new Home().Save(path, data).write();` where `Save` inits class fields `Path path` and `Input input`
Did I get the idea?

yegor256 replied:
@yasamprom no, this is wrong:

```
a(x) {
  f(Y(x));
}
a(y) {
  f(y);
}
```

this is right:

```
a(x) {
  a(Y(x));
}
a(y) {
  f(y);
}
```

yegor256 commented:
@yasamprom much better now! one comment above

yasamprom commented:
@yegor256 please check

yegor256 commented on eo-maven-plugin/src/main/java/org/eolang/maven/GmiMojo.java:
```diff
@@ -374,18 +374,18 @@ private void makeGraph(final String xembly, final Path gmi) throws IOException {
-                ).save();
+                new Home().save(
+                    gmi.resolveSibling(String.format("%s.dot", gmi.getFileName())),
+                    new Xsline(GmiMojo.TO_DOT).pass(graph).xpath("//dot/text()").get(0)
```
@yasamprom why did you copy it here? we already calculate `dot` a few lines above

yegor256 commented:
@yasamprom why didn't you do the same with `saveQuietly`? also, I would suggest to get rid of these methods at all

yasamprom commented:
> @yasamprom why didn't you do the same with `saveQuietly`? also, I would suggest to get rid of these methods at all

Do we want to remove all `saveQuietly` methods and replace them with save? Or remove only unused? I added 

> @yasamprom why didn't you do the same with `saveQuietly`? also, I would suggest to get rid of these methods at all

Ok, I'll remove these methods and replace with simple `save`

yasamprom commented:
@yegor256 Done

yegor256 commented:
@yasamprom see above
